paddle/fluid/framework/grad_op_desc_maker.h

```diff
@@ -132,22 +141,47 @@ class GradOpDescMakerBase {
   std::string ForwardOpType() const { return this->fwd_op_.Type(); }
 
  protected:
-  const OpDesc& ForwardOp() const { return fwd_op_; }
+  // const OpDesc& ForwardOp() const { return fwd_op_; }
+
+  bool HaveInput(const std::string& name) const {
+    if (var_base_in_ != nullptr && var_base_out_ != nullptr) {
+      // dygraph mode
+      auto it = var_base_in_->find(name);
+      return it != var_base_in_->end();
+    } else {
+      return (fwd_op_.Inputs().count(name) > 0);
+    }
+  }
 
  private:
   const OpDesc& fwd_op_;
   const std::unordered_set<std::string>& no_grad_set_;
   std::unordered_map<std::string, std::string>* grad_to_var_;
 
  protected:
+  const imperative::NameVarBaseMap* var_base_in_;
```

Comment:
Remove `var_base_in_` and `var_base_out_`.

Reply:
done